### PR TITLE
[CARBONDATA-3020] support lz4 as column compressor

### PR DESCRIPTION
**Pull request by kevinjmh**
lz4  looks not much difference than snappy.

### *Test Case 1*
Data generated by tpch-dbgen(lineitem)

### Loading Performance(compress)

create table with different 'carbon.column.compressor'

**scale factor=10 / 7.3G**

compressor | load time | file size after loading
-- | -- | --
snappy | 453.82s | 2.1G
zstd | 466.752s | 1.6G
lz4 | 443.373s | 2.1G

### Query Performance(decompress)

To force carbon to decompress the data, we create a spark project to `select * from TABLE` and iterate the result rdd. The total execution time is recorded for comparison.

**scale factor=10 / 7.3G**

compressor | full scan time (s) 
-- | -- 
snappy | 128.379 
zstd | 134.4814 
lz4 | 131.732 

### *Test Case 2*
Data generated by `org.apache.carbondata.spark.util.DataGenerator`

### Loading Performance(compress)

**100 millions records / 8.7G**

compressor | load time | file size after loading
-- | -- | --
snappy | 289.337s | 2.5G
zstd | 219.1s | 1.1G
lz4 | 285.072s | 2.3G

### Query Performance(decompress)

**100 millions records / 8.7G**

compressor | full scan time (s)
-- | -- 
snappy | 133.752 
zstd | 144.801 
lz4 | 134.938 

 

### *Reference*
Benchmark from https://github.com/lz4/lz4

|  Compressor             | Ratio   | Compression | Decompression |
|  ----------             | -----   | ----------- | ------------- |
|**LZ4 default (v1.8.2)** |**2.101**|**730 MB/s** | **3900 MB/s** |
|  Snappy 1.1.4           |  2.091  |  525 MB/s   |   1750 MB/s   |
|  [Zstandard] 1.3.4 -1   |  2.877  |  470 MB/s   |   1380 MB/s   |
|**LZ4 HC -9 (v1.8.2)**   |**2.721**|   40 MB/s   | **3920 MB/s** |

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8578/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/508/

File: core/src/main/java/net/jpountz/lz4/LZ4CompressorWithLength.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+/**
+ * Covenience class to include the length of the original decompressed data
+ * in the output compressed data, so that the user does not need to save
+ * the length at anywhere else.  The compressed data must be decompressed by
+ * {@link LZ4DecompressorWithLength} and is NOT compatible with any other
+ * decompressors in lz4-java or any other lz4 tools.  This class deliberately
+ * does not extend {@link LZ4Compressor} because they are not interchangable.
+ */
+
+public class LZ4CompressorWithLength {
```
Comment (jackylk) [inline core/src/main/java/net/jpountz/lz4/LZ4CompressorWithLength.java]: Is this copied from net.jpountz.lz4?

Comment (kevinjmh) [reply core/src/main/java/net/jpountz/lz4/LZ4CompressorWithLength.java]: yes.  These codes didn't not packed in a released jar. Here we cope it only for test.  See comment in L18-20

File: core/src/main/java/net/jpountz/lz4/LZ4DecompressorWithLength.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+  /**
+   * Convenience method, equivalent to calling
+   * {@link #decompress(byte[], int, byte[], int) decompress(src, 0, dest, 0)}.
+   *
+   * @param src the compressed data
+   * @param dest the destination buffer to store the decompressed data
+   * @return the number of bytes read to restore the original input
+   */
+  public int decompress(byte[] src, byte[] dest) {
```
Comment (xuchuanyin) [inline core/src/main/java/net/jpountz/lz4/LZ4DecompressorWithLength.java]: As we talked offline, the return value may not be the uncompressed size, we need to handle it.

Comment (kevinjmh) [reply core/src/main/java/net/jpountz/lz4/LZ4DecompressorWithLength.java]: ok, fixed

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/341/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8588/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/518/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/349/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/350/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/527/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8597/

**Comment by kevinjmh** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/360/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8607/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/537/

File: core/src/main/java/org/apache/carbondata/core/datastore/compression/Lz4Compressor.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+  @Override
+  public String getName() {
+    return "lz4";
+  }
+
+  @Override
+  public byte[] compressByte(byte[] unCompInput) {
+    // get max compressed length
```
Comment (xuchuanyin) [inline core/src/main/java/org/apache/carbondata/core/datastore/compression/Lz4Compressor.java]: If this code is copied from LZ4, better to make a declaration

**Comment by xuchuanyin** [comment]
better to provide performance test result for this

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/705/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8772/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/527/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/806/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9071/

File: core/src/main/java/org/apache/carbondata/core/datastore/compression/Lz4Compressor.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.Arrays;
+
+import org.apache.carbondata.core.util.ByteUtil;
+
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.lz4.LZ4FastDecompressor;
+
+
+public class Lz4Compressor implements Compressor {
```
[the remark this answers is not in the document]
Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/datastore/compression/Lz4Compressor.java]: added

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1096/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1307/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9359/

**Comment by kevinjmh** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1098/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9361/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1309/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1984/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2161/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/56/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/55/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/56/

**Pull request closed without merging**